File: _extensions/query-garden/resources/python/querygarden.py
from pathlib import Path


def _option_line(name, value):
    if value is None:
        return None
    if isinstance(value, bool):
        value = "true" if value else "false"
    return f"#| {name}: {value}"


def _fence(sql):
    fence = "```"
    while fence in sql:
        fence += "`"
    return fence


def sql_block(
    sql,
    *,
    db,
    sql_file_path=None,
    dialect=None,
    editable=None,
    output_location=None,
):
    """Return a Query Garden SQL block as markdown."""
    sql = str(sql).strip()
    fence = _fence(sql)
    options = [
        _option_line("query-garden", True),
        _option_line("db", db),
        _option_line("sql-file-path", sql_file_path),
        _option_line("editable", editable),
        _option_line("dialect", dialect),
        _option_line("output-location", output_location),
    ]
    option_text = "\n".join(option for option in options if option is not None)
    return f"{fence}{{.sql}}\n{option_text}\n{sql}\n{fence}"


def sql_block_from_file(query_file_path, **kwargs):
    """Read SQL from a file and return a Query Garden SQL block."""
    sql = Path(query_file_path).read_text(encoding="utf-8")
    return sql_block(sql, **kwargs)


def emit_sql_block(sql, **kwargs):
    """Print a Query Garden SQL block for Quarto cells using output: asis."""
    print(sql_block(sql, **kwargs))


def emit_sql_block_from_file(query_file_path, **kwargs):
    """Print a file-backed Query Garden SQL block for Quarto output: asis."""
    print(sql_block_from_file(query_file_path, **kwargs))

File: _extensions/query-garden/resources/r/querygarden.R
.query_garden_option_line <- function(name, value) {
  if (is.null(value)) {
    return(NULL)
  }
  if (is.logical(value)) {
    value <- if (isTRUE(value)) "true" else "false"
  }
  paste0("#| ", name, ": ", value)
}

.query_garden_fence <- function(sql) {
  fence <- "```"
  while (grepl(fence, sql, fixed = TRUE)) {
    fence <- paste0(fence, "`")
  }
  fence
}

query_garden_sql_block <- function(
  sql,
  db,
  sql_file_path = NULL,
  dialect = NULL,
  editable = NULL,
  output_location = NULL
) {
  sql <- trimws(paste(sql, collapse = "\n"))
  fence <- .query_garden_fence(sql)
  options <- c(
    .query_garden_option_line("query-garden", TRUE),
    .query_garden_option_line("db", db),
    .query_garden_option_line("sql-file-path", sql_file_path),
    .query_garden_option_line("editable", editable),
    .query_garden_option_line("dialect", dialect),
    .query_garden_option_line("output-location", output_location)
  )

  paste0(
    fence, "{.sql}\n",
    paste(options, collapse = "\n"), "\n",
    sql, "\n",
    fence
  )
}

query_garden_sql_block_from_file <- function(query_file_path, ...) {
  sql <- paste(readLines(query_file_path, warn = FALSE, encoding = "UTF-8"), collapse = "\n")
  query_garden_sql_block(sql, ...)
}

query_garden_emit_sql_block <- function(sql, ...) {
  cat(query_garden_sql_block(sql, ...), sep = "\n")
}

query_garden_emit_sql_block_from_file <- function(query_file_path, ...) {
  cat(query_garden_sql_block_from_file(query_file_path, ...), sep = "\n")
}

